Lesson3

ilya-burinskiy opened the pull request.

Commit:
TestsStartedByUser model

Commit:
Model seeds

Commit:
User tests_by_level instance method

Commit:
Test by_category class method

VidgarVii commented on app/models/test.rb:
```diff
@@ -1,2 +1,6 @@
+  def self.by_category(pattern)
+    tests_categories = Test.joins('INNER JOIN categories ON tests.category_id = categories.id')
+    tests_categories.where('categories.name LIKE :pattern', pattern: "%#{pattern}%").order('tests.title DESC')
+  end
```
1. Создайте метод класса в модели Test, который возвращает отсортированный по убыванию **массив названий всех Тестов** у которых Категория называется определённым образом (название категории передается в метод в качестве аргумента). 
2. Для оператора where ОРМ даёт более крутой синтаксис. where(categories: {title: category})
LIKE тут не очень оптимальный вариант
3 И order(title: :desc)

VidgarVii replied:
1 INNER писать не обязательно. Данный тип соединения по умолчанию стоит.
2 А где массив названий Тестов?

VidgarVii commented on app/models/user.rb:
```diff
@@ -1,2 +1,6 @@
+  def tests_by_level(level)
+    tests_ids = TestsStartedByUser.where(user_id: id).pluck(:test_id)
+    Test.where(id: tests_ids).where(level: level)
+  end
```
1 Создайте инстанс-метод в модели User, который принимает в качестве аргумента значение уровня сложности и возвращает список всех Тестов, которые **проходит или когда-либо проходил Пользователь** на этом уровне сложности.
Хорошо бы дать в названии метода указать что за тесты.
2 Тут тоже лучше использовать joins. И весь код может уложиться в 1 строчку. И будет всего 1 запрос в БД, а не 2

VidgarVii replied:
запрос не корректный. Нет условия у какого юзера показывать тесты

VidgarVii commented on db/seeds.rb:
```diff
@@ -5,3 +5,99 @@
+seed_categories
+seed_users(5)
+seed_tests
+seed_tests_started_by_user
```
0. Следи за отступами. 
1. Область видимости тут файл. Лучше обойтись без методов, а результаты записывать в переменные. Это позволит не лезть лишний раз в БД и не искать там данные для построения ассоциаций.
2 Сиды это начальные данные для разработки. Бывает так что нужно прогнать сид в процессе работы. По этому все данные должны быть защищены от дублирования. В данном случаи Если я запущу 10 db:seed, то у меня буде 30 категорий и и.д. Можно например проверять так User.exists?

ПС. Для прода тоже можно создавать сиды как стартовые данные, если код будут деплоить в 1 раз. Но это уже делается специальными гемами, где можно указать окружение

Commit:
Seed records are no longer duplicated

Commit:
WHERE query fix

Commit:
User tests by level method

- test_by_level renamed to started_test_by_level
- Using 1 query to database instead 2

Commit:
Category seed

Category seed record is saved when such category does not exist

VidgarVii commented on db/seeds.rb:
```diff
@@ -5,3 +5,114 @@
+    test_id: seed.test_id,
+    user_id: seed.user_id
+  ).nil? 
+end
```
1 Есть более подходящие для такого случая методы find_or_initialize/find_or_create
2 Нет пустой строки в конце файла

Commit:
Using find_or_create_by method

Commit:
Test by category method fix

Returning array of test titles

Commit:
User started tests by level method

Returning started tests of a given level for the user who called
started_tests_by_level method, not all tests with this level

Commit:
rubocop

VidgarVii commented on app/models/test.rb:
```diff
@@ -1,2 +1,7 @@
 class Test < ApplicationRecord
+  def self.by_category(pattern)
```
Отлично. теперь метод вернет то, что требовалось. А вот название метода не говорит об этом

Commit:
Test model fix

Renamed class method by_category to list_titles_by_category
